service-base/src/main/java/org/eclipse/hono/service/tenant/BaseTenantService.java

```diff
@@ -0,0 +1,305 @@
+                    log.debug("tenant request contains invalid structure");
+                    reply(tenantMsg, TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+                    break;
+                }
+                payload = getRequestPayload(body);
+                log.debug("creating tenant [{}] with data {}", tenantId, payload != null ? payload.encode() : null);
+                add(tenantId, payload, result -> reply(tenantMsg, result));
+                break;
+            case ACTION_UPDATE:
+                payload = getRequestPayload(body);
+                log.debug("updating tenant [{}] with data {}", tenantId, payload != null ? payload.encode() : null);
+                update(tenantId, payload, result -> reply(tenantMsg, result));
+                break;
+            case ACTION_REMOVE:
+                log.debug("deleting tenant [{}]", tenantId);
+                remove(tenantId, result -> reply(tenantMsg, result));
+                break;
+            default:
+                break;
+            }
+        } catch (final ClassCastException e) {
+            log.debug("malformed request message [{}]", e.getMessage());
+            reply(tenantMsg, TenantResult.from(HTTP_BAD_REQUEST));
+        }
+    }
+
+    private void reply(final Message<JsonObject> request, final AsyncResult<TenantResult> result) {
+        if (result.succeeded()) {
+            this.reply(request, result.result());
+        } else {
+            request.fail(HTTP_INTERNAL_ERROR, "cannot process tenant management request");
+        }
+    }
+
+    /**
+     * Sends a response to a tenant request over the Vertx event bus.
+     *
+     * @param request The message to respond to.
+     * @param result The tenant result that should be conveyed in the response.
+     */
+    protected final void reply(final Message<JsonObject> request, final TenantResult result) {
+        final JsonObject body = request.body();
+        final String tenantId = body.getString(RequestResponseApiConstants.FIELD_TENANT_ID);
+        request.reply(TenantConstants.getServiceReplyAsJson(tenantId, result));
+    }
+
+    private JsonObject getRequestPayload(final JsonObject request) {
+        final JsonObject payload = request.getJsonObject(TenantConstants.FIELD_PAYLOAD, new JsonObject());
+        addNotPresentFieldsWithDefaultValuesForTenant(payload);
+        return payload;
+    }
+
+    private boolean isValidRequestPayload(final JsonObject request) {
+        final JsonObject payload = request.getJsonObject(TenantConstants.FIELD_PAYLOAD, new JsonObject());
+        final JsonArray adapters = payload.getJsonArray(TenantConstants.FIELD_ADAPTERS);
+        if (adapters != null) {
+            if (adapters.size() == 0) {
+                return false;
+            }
+            for (int i = 0; i < adapters.size(); i++) {
+                final JsonObject adapterDetails = adapters.getJsonObject(i);
+                if (adapterDetails == null) {
```

Comment:
this cannot happen, but you can get a `ClassCastException` in the line above

Reply:
Will be disappearing with the stream-based solution.
The ClassCastException is caught in the calling code, but I will add javadoc to that.